applications/minotari_app_grpc/src/authentication/basic_auth.rs

```diff
@@ -156,26 +247,322 @@ mod tests {
     }
 
     mod validate {
+        use std::{
+            cmp::{max, min},
+            thread::sleep,
+        };
+
+        use rand::RngCore;
+        use tari_utilities::{hex::Hex, ByteArray};
+
         use super::*;
         use crate::authentication::salted_password::create_salted_hashed_password;
 
         #[test]
-        fn it_validates_for_matching_credentials() {
-            let hashed = create_salted_hashed_password(b"secret").unwrap();
-            let credentials = BasicAuthCredentials::new("admin".to_string(), hashed.to_string().into());
-            credentials.validate("admin", b"secret").unwrap();
+        fn it_validates_for_matching_salted_credentials() {
+            let hashed_password = create_salted_hashed_password(b"secret").unwrap();
+            // Typical username
+            let credentials =
+                BasicAuthCredentials::new("admin".to_string(), hashed_password.to_string().into()).unwrap();
+            credentials.constant_time_validate("admin", b"secret").unwrap();
+            // Empty username is also fine
+            let credentials = BasicAuthCredentials::new("".to_string(), hashed_password.to_string().into()).unwrap();
+            credentials.constant_time_validate("", b"secret").unwrap();
         }
 
         #[test]
-        fn it_rejects_for_mismatching_credentials() {
-            let credentials = BasicAuthCredentials::new("admin".to_string(), "bruteforce".to_string().into());
-            let err = credentials.validate("admin", b"secret").unwrap_err();
-            assert!(matches!(err, BasicAuthError::InvalidPassword(_)));
+        fn it_rejects_registering_unsalted_password_credentials() {
+            let err = BasicAuthCredentials::new("admin".to_string(), "secret".to_string().into()).unwrap_err();
+            assert!(matches!(
+                err,
+                BasicAuthError::InvalidPassword(argon2::password_hash::Error::PhcStringInvalid)
+            ));
+        }
 
-            let credentials = BasicAuthCredentials::new("bruteforce".to_string(), "secret".to_string().into());
-            let err = credentials.validate("admin", b"secret").unwrap_err();
+        #[test]
+        fn it_rejects_validating_mismatching_credentials() {
```

Comment:
For completeness, could you add a test with both incorrect username and passphrase? This will obviously (correctly) fail, but it seems prudent given the structure of the design (and could catch a future regression).